### PR TITLE
PPLiteSeg modules

### PR DESCRIPTION
**Pull request by lkdci**
Torch implementation of PPLiteSeg [paper](https://arxiv.org/abs/2204.02681).
Training recipes will be added later.

### CODE REVIEW

File: src/super_gradients/training/models/segmentation_models/ppliteseg.py [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+                 decoder_upsample_mode: Union[UpsampleMode, str],
+                 head_upsample_mode: Union[UpsampleMode, str],
+                 head_mid_channels: int,
+                 dropout: float,
+                 use_aux_heads: bool,
+                 aux_hidden_channels: List[int],
+                 aux_scale_factors: List[int]
+                 ):
+        """
+        :param backbone: Backbone nn.Module should implement the abstract class `AbstractSTDCBackbone`.
```
Comment (ofrimasad) [inline src/super_gradients/training/models/segmentation_models/ppliteseg.py:223]: looks like you are limiting the backbone to be an `AbstractSTDCBackbone` and not `nn.Module`

File: src/super_gradients/training/models/segmentation_models/ppliteseg.py [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+        :param projection_channels_list: channels list to project encoder features before fusing with the decoder
+            stream.
+        :param sppm_inter_channels: num channels in each sppm pooling branch.
+        :param sppm_out_channels: The number of output channels after sppm module.
+        :param sppm_pool_sizes: spatial output sizes of the pooled feature maps.
+        :param sppm_upsample_mode: Upsample mode to original size after pooling.
+        :param :
```
Comment (ofrimasad) [inline src/super_gradients/training/models/segmentation_models/ppliteseg.py:230]: please removeredundant lines

File: src/super_gradients/training/models/segmentation_models/ppliteseg.py [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+                                        upsample_mode=decoder_upsample_mode,
+                                        align_corners=align_corners)
+
+        # Init Segmentation classification heads
+        self.seg_head = nn.Sequential(
+            SegmentationHead(in_channels=decoder_channels[-1],
+                             mid_channels=head_mid_channels,
+                             num_classes=num_classes,
+                             dropout=dropout),
+            make_upsample_module(scale_factor=8, upsample_mode=head_upsample_mode, align_corners=align_corners)
```
Comment (ofrimasad) [inline src/super_gradients/training/models/segmentation_models/ppliteseg.py:264]: is it always 8? even if we define a different backbone?

File: src/super_gradients/training/utils/module_utils.py [hunk 4]
```diff
@@ -219,3 +220,30 @@ def __init__(self, mean_original, std_original, mean_required, std_required):
     def forward(self, x):
         x = (x + self.additive) * self.multiplier
         return x
+
+
+class UpsampleMode(Enum):
```
Comment (ofrimasad) [inline src/super_gradients/training/utils/module_utils.py:225]: please move this into a separate file under `src/super_gradients/common/data_types/enum`

**Review by ofrimasad — COMMENTED** [review]
really beautiful work!  
some minor comment inline

File: src/super_gradients/training/models/segmentation_models/ppliteseg.py [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+from typing import Union, List, Tuple
+from super_gradients.training.utils.module_utils import ConvBNReLU, UpsampleMode, make_upsample_module
+from super_gradients.training.models.segmentation_models.stdc import SegmentationHead, AbstractSTDCBackbone, STDC1Backbone, STDC2Backbone
+from super_gradients.training.models.segmentation_models.segmentation_module import SegmentationModule
+from super_gradients.training.utils import HpmStruct, get_param
+
+
+class SPPM(nn.Module):
+    """
+    Simple Pyramid Pooling context Module.
```
[the remark this answers is not in the document]
Comment (lkdci) [reply src/super_gradients/training/models/segmentation_models/ppliteseg.py:13]: It is actually a design of the ppliteseg paper authors, they are similar in a way of applying pooling layers with different kernel sizes. The differences are that here average pooling is used instead of max pooling and the branches are summed instead concatenation.

File: src/super_gradients/training/models/segmentation_models/ppliteseg.py [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+            ) for pool_size in pool_sizes
+        ])
+        self.conv_out = ConvBNReLU(inter_channels, out_channels, kernel_size=3, padding=1, bias=False)
+        self.out_channels = out_channels
+        self.upsample_mode = upsample_mode
+        self.align_corners = align_corners
+        self.pool_sizes = pool_sizes
+
+    def output_channels(self):
+        return self.out_channels
```
[the remark this answers is not in the document]
Comment (lkdci) [reply src/super_gradients/training/models/segmentation_models/ppliteseg.py]: Since SPPM is hard-coded I don't really need this specific method. I was probably thinking of a future structure and factory for context modules.
But it is a work for a different PR.
Remove `out_channels()` method, simply call `sppm.out_channels` [bd37f20](https://github.com/Deci-AI/super-gradients/pull/349/commits/bd37f20eb277d8bd5ae6a0b33b1fbbd5a3315c08)

File: src/super_gradients/training/models/segmentation_models/ppliteseg.py [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+            global_pool: nn.AdaptiveAvgPool2d = branch[0]
+            out_size = global_pool.output_size
+            out_size = out_size if isinstance(out_size, (tuple, list)) else (out_size, out_size)
+            kernel_size = [int(i / o) for i, o in zip(input_size, out_size)]
+            branch[0] = nn.AvgPool2d(kernel_size=kernel_size, stride=kernel_size)
+
+
+class UAFM(nn.Module):
+    """
+    Unified Attention Fusion Module, which uses mean and max values across the spatial dimensions.
```
[the remark this answers is not in the document]
Comment (lkdci) [reply src/super_gradients/training/models/segmentation_models/ppliteseg.py:66]: Already had under `PPLiteSegBase`. Do you think the head of the file is a better place @ofrimasad ?
```
class PPLiteSegBase(SegmentationModule):
    """
    The PP_LiteSeg implementation based on PaddlePaddle.
    The original article refers to "Juncai Peng, Yi Liu, Shiyu Tang, Yuying Hao, Lutao Chu,
    Guowei Chen, Zewu Wu, Zeyu Chen, Zhiliang Yu, Yuning Du, Qingqing Dang,Baohua Lai,
    Qiwen Liu, Xiaoguang Hu, Dianhai Yu, Yanjun Ma. PP-LiteSeg: A Superior Real-Time Semantic
    Segmentation Model. https://arxiv.org/abs/2204.02681".
    """
    def __init__(self,
```

File: src/super_gradients/training/models/segmentation_models/ppliteseg.py (hunk 1 above)
Comment (lkdci) [reply src/super_gradients/training/models/segmentation_models/ppliteseg.py:220]: Because I need the same constraints, and abstract method `get_backbone_output_number_of_channels` to be able to extract the backbone num_channels, in order to build the decoder.

File: src/super_gradients/training/models/segmentation_models/ppliteseg.py (hunk 2 above)
Comment (lkdci) [reply src/super_gradients/training/models/segmentation_models/ppliteseg.py]: It was actually placeholders for rest of args doc, added. [c7ab444](https://github.com/Deci-AI/super-gradients/pull/349/commits/c7ab4449a7723dd379d829a9d2ebca068e4d5b0b)

File: src/super_gradients/training/models/segmentation_models/ppliteseg.py (hunk 3 above)
Comment (lkdci) [reply src/super_gradients/training/models/segmentation_models/ppliteseg.py]: Added a scale_factor arg in case a custom design is required. [2972d7a](https://github.com/Deci-AI/super-gradients/pull/349/commits/2972d7aa52b156c2fafca28a4437ab7b63c8cc8b)

File: src/super_gradients/training/utils/module_utils.py (hunk 4 above)
Comment (lkdci) [reply src/super_gradients/training/utils/module_utils.py]: Done [8302bdf](https://github.com/Deci-AI/super-gradients/pull/349/commits/8302bdfb3d68f0fab4d04ba201d122f0184c0443)

**Comment by lkdci** [comment]
Addressed @ofrimasad

**Review by ofrimasad — APPROVED** [review]
LGTM
